### ISSUE TITLE
Toggle plugin broken

### ISSUE THREAD

**Issue by vallieresc** [wet-boew/wet-boew#6471]
I've been trying to implement your simple toggle plugin using a button and a simple div but I can't seem to get it to work for some reason. I have no js errors, no typos. I see the on/off class toggle on the button but my help div is not toggling. I also noticed your simple toggle example is not working. I am running WET 4.0.9.

My generated button: 

```html
<button type="button" class="wb-toggle wb-toggle-inited on" data-toggle="{'selector': '.alert-warning'}" id="2dd85e5d-1139-4f83-999c-8bf33a6afaaa" aria-controls="2dd85e5d-1139-4f83-999c-8bf33a6afaaa">Toggle</button>
```

My div:
```html
<div class="title-block"><div class="title-help alert alert-warning">Example of help text.</div></div>
```

**Comment by pjackson28** [comment]
Simple toggle example is working for me on the live working examples (http://wet-boew.github.io/v4.0-ci/demos/toggle/toggle-en.html). Does it work for you there? If so, perhaps there is an issue with v4.0.9 (if so then updating to v4.0.10 should fix it).

**Comment by vallieresc** [comment]
The simple toggle example on that page doesn't seem to work in chrome but it does in firefox. I still can't toggle in my app. Have you made any fixes to the toggle plugin in 4.0.10? I've made some small changes in wet-boew.js.

**Comment by pjackson28** [comment]
The live working example works fine for me in Chrome and Firefox. It is possible that changes were made between v4.0.9 and v4.0.10. 

As for changing wet-boew.js, you shouldn't be making any changes that file as we don't support custom modifications to that file. If there are changes that are needed then they should be proposed through a PR so they can be included in the central code. If you need custom JS then it should be kept to a separate file.

**Comment by vallieresc** [comment]
Ok I'll update to 4.0.10 and make the PR wet-boew.js changes are still necessary.

**Comment by pjackson28** [comment]
What are you changing in wet-boew.js and why is it necessary? Perhaps we can help to find an alternate solution that doesn't involve changes to wet-boew.js as changing wet-boew.js creates issues going forward (maintenance/upgrade headaches, problems if you move to a platform where wet-boew.js is locked down, etc.). Also note that you should be using the minified version of wet-boew.js for performance/download size reasons (which then makes modifications to wet-boew.js even more difficult to support/maintain).

**Comment by vallieresc** [comment]
I modified web-boew.js because in our left menu navigation we have non linked items that should be included in the mobile panel. Without my changes these items cause a nodeName js error. They are <li><span>Link</span></li>. Also in our footer we do not have any link below the h3 so I've had to change the jquery selector to include all a tags in the footer. 

On save my changes are reminified.

**Comment by pjackson28** [comment]
You should be filing issues about these rather than modifying wet-boew.js. In those issues you should be identifying the use case and the markup you are using so we can see if there is a better way to do the markup or if a change in core is needed.

**Comment by pjackson28** [comment]
Also if you have suggested code changes then you are welcome to submit a PR and we can review your suggested changes.

**Comment by vallieresc** [comment]
Noted.

The getElements = function( link, data ) {
		var selector = data.selector || link,
			parent = data.parent || null;

		return parent !== null ? $( parent ).find( selector ) : $( selector );
	},

seems to return the button itself. Data.selector is undefined and link is the button.

**Comment by pjackson28** [comment]
There are some odd things with your markup. For instance the <code>id</code> and <code>aria-controls</code> attributes on the button are the same meaning that the button controls itself (where instead <code>aria-controls</code> should be the same as the <code>id</code> on the toggled element). This doesn't happen in the live working examples. What is the original markup that you are using (so not the generated markup)?

**Comment by vallieresc** [comment]
@pjackson28 The id and aria controls was happening because the plugin couldn't pickup the selector.

**Comment by pjackson28** [comment]
That is expected. The first example is invalid JSON while the second is valid JSON. WET requires valid JSON to be passed through the data-* attributes and ignores anything that is not valid (for security reasons). So when you used invalid JSON, it was the same as if you specified no selector at all (hence why the button referenced itself).

**Comment by neoinsight** [comment]
I share @pjackson28's concern about modifying the .js file from a usability point of view. WET menus force an error for unlinked menu items because well, it's a menu, there are supposed to be links. 
When we perform usability testing, we use the .js version of WET so if you are modifying, it's also a usability risk.

**Comment by vallieresc** [comment]
@neoinsight I agree but our complex application has a multi tier navigation like a tree. In some cases a non link element is used to group sub pages so users can better understand the navigation.

**Comment by neoinsight** [comment]
Uh oh. 'complex application with multi tier navigation like a tree'. From a web navigation point of view, it's always preferable to group and navigate via linked elements, even if you have to add some simple hub pages (like these - sorry, don't have a WET4 example yet: http://www.dfo-mpo.gc.ca/otw-am/conditions/index-eng.htm) or like the whole Theme-Topic-Subtopic pages on Canada.ca. But agree it's tricky... Good luck.

**Comment by pjackson28** [comment]
@vallieresc Don't know how good it is from a usability standpoint but another option could be to turn the unlinked menu items into same-page links (so linking to each sub-menu). That would avoid the need for any JavaScript changes and would ensure the item is included in the tabbing order. That is what was done here: http://www.healthycanadians.gc.ca/eating-nutrition/safety-salubrite/index-eng.php. Also avoids the scenario where the user clicks on the unlinked item and it does nothing (where they expected a side menu item to do something when clicked). At least if it is a same-page link it does something when clicked (shifts focus to the sub-menu).

**Comment by vallieresc** [comment]
@neoinsight Sounds nice maybe in some cases but let's say I have 5 links under a non linked item. The non linked item would navigate to a sitemap page with the same 5 links. Seems a waste of a click to me and a lot of work for us.

@pjackson28 Could be a solution. What about the footer link lists? We do not have any lists. The h3 are the links and with your jquery selector .not('h3 a') it didn't work.

**Comment by pjackson28** [comment]
@vallieresc The problem is with your markup, as you should not be using an <code>h3</code> or <code>section</code> without section content (will fail WCAG 2.0). So if there are no sub-links, then remove the <code>section</code> element and change the <code>h3</code> element to a <code>div class="h3"</code> element (or something similar that doesn't involve <code>h3</code> or <code>section</code>).

Note the JS code you are having trouble with is:
```js
$info.find( "h3, a" ).not( "section a" )
```
It is intentionally filtering out any sub-links (and WCAG 2.0 violations) with the <code>.not( "section a" )</code>.

**Comment by neoinsight** [comment]
@pjackson28 That on-page solution seems like it would meet usability and as you point out, accessibility requirements -  clicking on items that do nothing was what worried me, as well as general menu clutter/overloading.

**Comment by vallieresc** [comment]
@pjackson28 For some reason clicking on my new anchored left menu item (links to submenu) will not visually navigate to the submenu. If I disable JS it will. Do you have any sort of preventDefault() or return false related to left menu anchors?

**Comment by pjackson28** [comment]
It works fine for me. What markup are you using for your left menu? Does the href match the id?

**Comment by vallieresc** [comment]
yes they match

```
<li><a href="#leftmenu-group0" class="list-group-item">Examples</a><ul id="leftmenu-group0" class="list-group menu list-unstyled"><li><a class="list-group-item status-completed wb-navcurr" href="/V001?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Simple Scenarios</a></li><li><a class="list-group-item" href="/V002?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Complex Scenarios</a></li><li><a class="list-group-item" href="/V003?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Supported Validations</a></li><li><a class="list-group-item" href="/V004?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">ReadOnly Controls</a></li></ul></li>
```

**Comment by pjackson28** [comment]
Your markup works for me when I insert it into a left side menu. Could you please provide the markup for your whole left side menu?

**Comment by vallieresc** [comment]
If I remove the web-boew.js script, the anchor works. I'm on 4.0.10.

```
<nav role="navigation" id="wb-sec" typeof="SiteNavigationElement" class="col-md-3 col-md-pull-9 visible-md visible-lg"><h2>Section menu</h2><ul class="list-group menu list-unstyled"><li><h3>Reporting Details</h3><ul class="list-group menu list-unstyled"><li><a href="#leftmenu-group0" class="list-group-item">Examples</a><ul id="leftmenu-group0" class="list-group menu list-unstyled"><li><a class="list-group-item status-completed wb-navcurr" href="/V001?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Simple Scenarios</a></li><li><a class="list-group-item" href="/V002?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Complex Scenarios</a></li><li><a class="list-group-item" href="/V003?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Supported Validations</a></li><li><a class="list-group-item" href="/V004?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">ReadOnly Controls</a></li></ul></li></ul></li></ul></nav>
```

**Comment by pjackson28** [comment]
The link is working, but because of how you are nesting your menu items the id is being included in the mobile panel because the nested markup is being brought over. This means you have duplicate id values with the one hidden in the mobile panel being earlier in the DOM. So that means on the click it is trying to set focus on the element hidden in the mobile panel which instead results in focus being set to body.

Perhaps we'll need to implement id filtering into the mobile panel generation. Will also need to see how your nested markup is being handled.

**Comment by pjackson28** [comment]
Part of the problem is you are going three levels deep in the left menu whereas WET only supports two levels deep. Your code even looks odd visually in the left menu as the Examples link is at the same level visually as the nested links:
![sidemenu](https://cloud.githubusercontent.com/assets/1326498/5939762/776241d8-a6d4-11e4-834a-2546f2990619.png)

It also causes issues in the mobile panel:
![sidemobile](https://cloud.githubusercontent.com/assets/1326498/5939841/05035e50-a6d5-11e4-84fb-6e2ffb090370.png)

Why do you have so much nesting in the side menu? Is the h3 even necessary? Everything works fine (including the link click and all visuals) if you drop the h3 and top level ul/li like the following:
```html
<nav role="navigation" id="wb-sec" typeof="SiteNavigationElement" class="col-md-3 col-md-pull-9 visible-md visible-lg">
<h2>Section menu</h2>
    <ul class="list-group menu list-unstyled">
		<li><a href="#leftmenu-group0" class="list-group-item">Examples</a>
			<ul id="leftmenu-group0" class="list-group menu list-unstyled">
				<li><a class="list-group-item status-completed wb-navcurr" href="/V001?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Simple Scenarios</a></li>
				<li><a class="list-group-item" href="/V002?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Complex Scenarios</a></li>
				<li><a class="list-group-item" href="/V003?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">Supported Validations</a></li>
				<li><a class="list-group-item" href="/V004?LZ80%2ft0Oud8EXERUojdprSmFQ7WhFYA8YOqnSMEFiJI%3d">ReadOnly Controls</a></li>
			</ul>
		</li>
    </ul>
</nav>
```
The two-level code is consistent with what is being done on HealthyCanadians.gc.ca.

**Comment by pjackson28** [comment]
Note you could use the h3 approach as well (like in http://wet-boew.github.io/themes-dist/theme-gcwu-fegc/content-secmenu-en.html) but you shouldn't be going past two levels deep (so don't combine both approaches into one).

**Comment by vallieresc** [comment]
This is only the demo project left menu. Other apps have more complex left menus. We cannot only support 2 levels. I take care of the mb-pnl layout with my css.

**Comment by pjackson28** [comment]
You may be fixing it with CSS but you are not fixing the accessibility issues, particularly the breaking of the WAI-ARIA menu pattern in the mobile panel (so there will be keyboard and screen reader issues). What value is the h3 bringing in this case?

Also why do you need to display all the options at once? Shouldn't the side menu be more contextual (so top tasks/actions for the current section/area)?

**Comment by LaurentGoderre** [comment]
Using a left menu for navigation seems very problematic to me. The navigation would probably best suited inside the content area.

**Comment by vallieresc** [comment]
Applications using our framework are made up of many form pages that users must fill out. It varies from 5 pages to 100. Some have left menus split in areas (only showing part of the menu). The left menu is the most important part of navigation for us. The left menu pages also show their completion status as well as facilitate their navigation to any page they need to fill out. The h3 could become a normal li span. Worst case I can generate the WAI-ARIA myself.

**Comment by pjackson28** [comment]
From what you describe it seems you are misusing the left menu. It is only intended for secondary navigation, it should not be relied upon for the primary (or the only) means of navigating and it should not contain anything that is not navigation (since it is in a <code>nav</code> element). Putting completion status in a nav element is against the HTML5 spec and makes no sense semantically (and may very well run afoul of WCAG 2.0). The completion status needs to be in the content area so it is visible by default across all views.

So right off the bat, anything non-navigation needs to be moved to the content area. It should not exist in the left menu. Use the grid system to position it appropriately in the content area.

Secondly, you should not be providing any navigation links exclusively in the side menu, particularly if it is important navigation. The links relevant to the page should be made available in the content area (where it makes the most sense contextually). You can use the grid system to position links appropriately if you need to (make sure it makes semantic sense). You can also have those same links in the side menu but they should not be in there exclusively.

You may not even need a side menu after such changes, and may very well benefit from the extra real-estate that you gain (which quite often is the case with complex applications where page space is at a premium).

**Comment by pjackson28** [comment]
Note also that with the grid system you could position some items on the left of the content area if that makes the most sense for your application (make sure to take into account how things should be displayed on smaller screens). You should not be using a side menu to just position stuff on the left much like you shouldn't be using a table solely for layout purposes. You should only use the side menu for it's intended purpose, which is to provide secondary navigation. If you want to put completion status on the left (not sure if that is a good idea from a usability standpoint), then use the grid system to do so.

**Comment by vallieresc** [comment]
When you say putting the left menu items in the content area what do you mean? I need those menu items available on all pages to facilitate navigation. Do you have an example somewhere of what you are suggesting?

**Comment by LaurentGoderre** [comment]
A progress status is not a navigation item though, it should be in the content area. Besides, it becomes quite useless to have it there in mobile view since it's hidden away

**Comment by pjackson28** [comment]
Anything that is not navigation moves to the content area. If you need to repeat those items across numerous pages then create an area in your content area grid for that purpose.

As for navigation, focus should be on the content area. Identify what are the most common next steps and provide that through the content area (what makes the most sense in context). The user should be able to navigate through your application without ever seeing the side menu (which could very well be the case with some mobile users). If you choose to have a side menu then put the relevant links as well but they should be considered a secondary means of navigating and not what users should rely on (since they are hidden by default on mobile and can't benefit from context like in the content area). The side menu should not go more than two levels deep as otherwise you are throwing too much at the user (focus on what is relevant).

**Comment by vallieresc** [comment]
I'm not sure what type of menu you mean. Do you mean the type of menu on this page http://wet-boew.github.io/v4.0-ci/demos/toggle/toggle-en.html ?

**Comment by pjackson28** [comment]
The side menu (the "Section menu") is on the left side of the page in desktop view (http://wet-boew.github.io/themes-dist/theme-gcwu-fegc/content-secmenu-en.html). In mobile view (shrink viewport below 992px), the side menu gets hidden and the mobile panel button gets revealed (the button on the top right with the magnifying glass and the lines). In mobile view the "Section menu" is included in the mobile panel.

So if you have any critical links that need to be visible by default in all cases, don't put them in the side menu as they will not be visible by default in mobile view.

**Comment by pjackson28** [comment]
Note that PR #6471 adds more tolerance for nested menu items but you still need to revisit how you are using side menus and ensure that they are used properly (or switch to an alternate approach such as with the grid system).

**Comment by vallieresc** [comment]
What would be the best way to have an css class toggle with the state of the toggle target but only after it has toggled? Basically I want to have a glyphicon arrow indicate that the element is open with a down and up arrow. If I have a click event, if executes before the plugin, the .on/.off class has not been updated yet.

**Comment by pjackson28** [comment]
For opening and closing elements you should be using HTML5 details/summary as that is what those elements are intended for (and the triangle is taken care of for you). You should be consistent with the convention for opened/closed that is already in use and try to avoid implementing different conventions.

**Comment by vallieresc** [comment]
Can the summary replace an <hx> tag? We need collapsible sections (large chunk of form).

**Comment by pjackson28** [comment]
You can include such a tag in the summary element, just make sure to use the equivalent hx class on the summary element. For example:

```html
<details>
.	<summary class="h3"><h3>Section title</h3></summary>
	<form ...>
	...
	</form>
</details>
```

**Comment by vallieresc** [comment]
I have found that if an element has no state, which makes it visible on screen, and no type is passed, which means it toggles, the first click of the button will assume it is in off state when it is really in on state. Your first click will do nothing, only your second click will correctly toggle the state. The state when in a toggle mode and for a non details element should be taken from the visibility jquery flag.

**Comment by pjackson28** [comment]
The issue here is the plugin assumes that the initial state is off. There is a global <code>states</code> object associated to the plugin and it doesn't get populated until the first toggle. So on the first click of a stateless button (toggle button), it doesn't find any state information in the <code>states</code> object for the selector so just default to off. So the likely solution is to have the plugin determine what is the initial state on init and to populate the states object accordingly (so perhaps fire a setState on init).

**Comment by vallieresc** [comment]
That sounds like a good solution. Will this be part of the next version?

**Comment by pjackson28** [comment]
Will be fixed by #6495. You will be able to set the initial state through the <code>data-wb-toggle</code> attribute using the <code>state</code> property (either <code>on</code> or <code>off</code>).

**Comment by vallieresc** [comment]
Also the toggle plugin will not work if you have a span element inside the button tag like this:
```
<button type="button" class="btn btn-primary wb-toggle" data-toggle='{"selector": ".box"}'><span class="glyphicon glyphicon-tag"></span> Toggle</button>
```

**Comment by pjackson28** [comment]
That will be fixed by the PR as well.

**Comment by vallieresc** [comment]
Discovered a bug with the new 'state' property. If you set state to 'on' and have persist set to 'session', the toggle will not open on the first click after previously having closed the toggle before loading the page. The default state should be overridden by the persist state.

**Comment by LaurentGoderre** [comment]
@vallieresc can you please open a new issue for this bug?

**Comment by pjackson28** [comment]
@valllieresc Also please provide the markup you are using in that new issue. I wasn't able to reproduce the issue with the following:
```html
<details>
	<summary class="wb-toggle" data-toggle='{"persist": "session", "state": "on"}'>Example 1</summary>
	<p>Example content that provides more details.</p>
	<table>
		<caption>Cups of coffee consumed by each delegate</caption>
		<thead>
			<tr>
				<th scope="col">Name</th>
				<th scope="col">Cups</th>
				<th scope="col">Type of Coffee</th>
				<th scope="col">Sugar?</th>
			</tr>
		</thead>
		<tbody>
			<tr>
				<td>T. Sexton</td>
				<td>10</td>
				<td>Espresso</td>
				<td>No</td>
			</tr>
			<tr>
				<td>J. Dinnen</td>
				<td>5</td>
				<td>Decaf</td>
				<td>Yes</td>
			</tr>
		</tbody>
	</table>
</details>
```

**Comment by pjackson28** [comment]
Note that PR #6714 should eliminate any possibility of this happening. It makes it ignore the "state" option if there is a persistent state already.
